### ISSUE TITLE
RuntimeError: found negative index -562949566989948 for dim 1

### ISSUE THREAD

**Issue by zhongshsh** [enoche/MMRec#10]
when running DRAGON on Elec., `knn_ind` is abnormal whose value is shown below: 
```
tensor([[                 0,                  0,                  0,
          ...,                  0,                  0,
                          0],
        [                 0,    133148282176512,                  0,
          ...,                  0, 544373700175071744,
           3978033069303296],
        [283817417514181632,   4470614405284095,   4470614278539776,
          ...,     96757024170496,   4472813428539647,
           4636747909554688],
        ...,
        [             62998,               1757,               3143,
          ...,              23576,              34277,
                      37654],
        [             62999,              13636,              20836,
          ...,              28260,              33679,
                      59455],
        [             63000,              61593,              54526,
          ...,              10903,              60347,
                      57793]], device='cuda:0')
```

The error message is below:
```
gpu_id=0
use_gpu=True
seed=[999]
data_path=../data/
inter_splitting_label=x_label
filter_out_cod_start_users=True
is_multimodal_model=True
checkpoint_dir=saved
save_recommended_topk=True
recommend_topk=recommend_topk/
embedding_size=64
weight_decay=0.0
req_training=True
epochs=1000
stopping_step=20
train_batch_size=2048
learner=adam
learning_rate=[0.001]
learning_rate_scheduler=[1.0, 50]
eval_step=1
training_neg_sample_num=1
use_neg_sampling=True
use_full_sampling=False
NEG_PREFIX=neg__
USER_ID_FIELD=userID
ITEM_ID_FIELD=itemID
TIME_FIELD=timestamp
field_separator=
metrics=['Recall', 'NDCG', 'Precision', 'MAP']
topk=[5, 10, 20, 50]
valid_metric=Recall@20
eval_batch_size=4096
use_raw_features=False
max_txt_len=32
max_img_size=256
vocab_size=30522
type_vocab_size=2
hidden_size=4
pad_token_id=0
max_position_embeddings=512
layer_norm_eps=1e-12
hidden_dropout_prob=0.1
end2end=False
hyper_parameters=['aggr_mode', 'reg_weight', 'learning_rate', 'seed']
inter_file_name=elec.inter
vision_feature_file=image_feat.npy
text_feature_file=text_feat.npy
user_graph_dict_file=user_graph_dict.npy
feat_embed_dim=64
n_mm_layers=1
n_layers=2
knn_k=10
mm_image_weight=0.1
aggr_mode=['add']
reg_weight=[0.0001]
info=normal-seed_None
model=DRAGON
dataset=elec
valid_metric_bigger=True
device=cuda


19 Jul 16:13    INFO elec
The number of users: 192403
Average actions of users: 8.779426516218562
The number of items: 63001
Average actions of items: 26.81208234789924
The number of inters: 1689188
The sparsity of the dataset: 99.98606462355167%
19 Jul 16:13    INFO 
====Training====
elec
The number of users: 192403
Average actions of users: 6.519861956414401
The number of items: 62989
Average actions of items: 19.91523916874375
The number of inters: 1254441
The sparsity of the dataset: 99.98964920548602%
19 Jul 16:13    INFO 
====Validation====
elec
The number of users: 192403
Average actions of users: 1.098194934590417
The number of items: 47923
Average actions of items: 4.4090728877574445
The number of inters: 211296
The sparsity of the dataset: 99.99770841780649%
19 Jul 16:13    INFO 
====Testing====
elec
The number of users: 192403
Average actions of users: 1.161369625213744
The number of items: 48794
Average actions of items: 4.57947698487519
The number of inters: 223451
The sparsity of the dataset: 99.99761985156943%
19 Jul 16:14    INFO 

=================================


19 Jul 16:14    INFO =========1/1: Parameters:['aggr_mode', 'reg_weight', 'learning_rate', 'seed']=('add', 0.0001, 0.001, 999)=======
Traceback (most recent call last):
  File "main.py", line 34, in <module>
    quick_start(model=args.model, dataset=args.dataset, config_dict=config_dict, save_model=True, seed=args.seed, alpha1=args.alpha1, alpha2=args.alpha2, eta=args.eta, estart=args.estart, alter=True if 'alter' in args.info else False)
  File "xxx/MMRec/src/utils/quick_start.py", line 77, in quick_start
    model = get_model(config['model'])(config, train_data).to(config['device'])
  File "xxx/MMRec/src/models/dragon.py", line 75, in __init__
    indices, image_adj = self.get_knn_adj_mat(self.image_embedding.weight.detach())
  File "xxx/MMRec/src/models/dragon.py", line 173, in get_knn_adj_mat
    return indices, self.compute_normalized_laplacian(indices, adj_size)
  File "xxx/MMRec/src/models/dragon.py", line 176, in compute_normalized_laplacian
    adj = torch.sparse.FloatTensor(indices, torch.ones_like(indices[0]), adj_size)
RuntimeError: found negative index -562949566989948 for dim 1
```

**Comment by enoche** [comment]
Thanks for reporting the error.
We are freeing GPU resources to reproduce your reported error. We will identify the bug soon...

**Comment by hongyurain** [comment]
Hi, we have rerun DRAGON on Elec. The ```knn_ind``` is shown below: 
```
tensor([[    0, 37655, 39325,  ..., 59776, 55434, 37525],
        [    1, 60006,  7130,  ..., 23531, 59370, 42955],
        [    2, 22890, 28742,  ..., 34440, 40991, 34408],
        ...,
        [62998,  1757,  3143,  ..., 23576, 34274, 34277],
        [62999, 13636, 20836,  ..., 28260, 33679, 59455],
        [63000, 61593, 54526,  ..., 10903, 60347, 57793]], device='cuda:0')
tensor([[    0, 37525, 37655,  ..., 20970, 39325,  8500],
        [    1,    14,     2,  ..., 55699, 41838, 40673],
        [    2,    14,     1,  ..., 56491, 21550, 56590],
        ...,
        [62998, 58360, 62223,  ..., 27326, 44843, 24995],
        [62999, 50218, 50219,  ...,  1903, 10708,  3288],
        [63000, 61593, 53392,  ..., 44834, 54602, 36302]], device='cuda:0')

```
The according model parameter settings are below:
```
gpu_id=0
use_gpu=True
seed=[999]
data_path=../data/
inter_splitting_label=x_label
filter_out_cod_start_users=True
is_multimodal_model=True
checkpoint_dir=saved
save_recommended_topk=True
recommend_topk=recommend_topk/
embedding_size=64
epochs=1000
stopping_step=20
train_batch_size=2048
learner=adam
learning_rate=[0.001]
learning_rate_scheduler=[1.0, 50]
eval_step=1
training_neg_sample_num=1
use_neg_sampling=True
use_full_sampling=False
NEG_PREFIX=neg__
USER_ID_FIELD=userID
ITEM_ID_FIELD=itemID
TIME_FIELD=timestamp
field_separator=
metrics=['Recall', 'NDCG', 'Precision', 'MAP', 'MRR']
topk=[5, 10, 20, 50]
valid_metric=Recall@20
eval_batch_size=4096
use_raw_features=False
max_txt_len=32
max_img_size=256
vocab_size=30522
type_vocab_size=2
hidden_size=4
pad_token_id=0
max_position_embeddings=512
layer_norm_eps=1e-12
hidden_dropout_prob=0.1
end2end=False
hyper_parameters=['aggr_mode', 'reg_weight', 'learning_rate', 'seed']
load_meta_cols=['itemID', 'description', 'title', 'category']
TEXT_ID_FIELD=description
inter_file_name=elec-indexed-v4.inter
text_file_name=meta-elec-indexed.csv
img_dir_name=img
vision_feature_file=image_feat.npy
text_feature_file=text_feat-v1.npy
user_graph_dict_file=user_graph_dict.npy
feat_embed_dim=64
n_mm_layers=1
n_layers=2
knn_k=10
mm_image_weight=0.1
aggr_mode=['add']
reg_weight=[0.0001]
u_layers=[1]
model=DRAGON
dataset=elec
valid_metric_bigger=True
device=cuda


21 Jul 10:30    INFO elec
The number of users: 192403
Average actions of users: 8.779426516218562
The number of items: 63001
Average actions of items: 26.81208234789924
The number of inters: 1689188
The sparsity of the dataset: 99.98606462355167%
21 Jul 10:30    INFO
====Training====
elec
The number of users: 192403
Average actions of users: 6.519861956414401
The number of items: 62989
Average actions of items: 19.91523916874375
The number of inters: 1254441
The sparsity of the dataset: 99.98964920548602%
21 Jul 10:30    INFO
====Validation====
elec
The number of users: 192403
Average actions of users: 1.098194934590417
The number of items: 47923
Average actions of items: 4.4090728877574445
The number of inters: 211296
The sparsity of the dataset: 99.99770841780649%
21 Jul 10:30    INFO
====Testing====
elec
The number of users: 192403
Average actions of users: 1.161369625213744
The number of items: 48794
Average actions of items: 4.57947698487519
The number of inters: 223451
The sparsity of the dataset: 99.99761985156943%
21 Jul 10:31    INFO

=================================


21 Jul 10:31    INFO =========1/1: Parameters:['aggr_mode', 'reg_weight', 'learning_rate', 'seed']=('add', 0.0001, 0.001, 999)=======
21 Jul 10:32    INFO DRAGON(
  (MLP_v): Linear(in_features=64, out_features=64, bias=False)
  (MLP_t): Linear(in_features=64, out_features=64, bias=False)
  (image_embedding): Embedding(63001, 4096)
  (image_trs): Linear(in_features=4096, out_features=64, bias=True)
  (text_embedding): Embedding(63001, 384)
  (text_trs): Linear(in_features=384, out_features=64, bias=True)
  (MLP_user): Linear(in_features=128, out_features=64, bias=True)
  (v_gcn): GCN(
    (MLP): Linear(in_features=4096, out_features=256, bias=True)
    (MLP_1): Linear(in_features=256, out_features=64, bias=True)
    (conv_embed_1): Base_gcn(64, 64)
  )
  (t_gcn): GCN(
    (MLP): Linear(in_features=384, out_features=256, bias=True)
    (MLP_1): Linear(in_features=256, out_features=64, bias=True)
    (conv_embed_1): Base_gcn(64, 64)
  )
  (user_graph): User_Graph_sample()
)
Trainable parameters: 308866456
21 Jul 10:39    INFO epoch 0 training [time: 459.09s, train loss: 150.7909]
21 Jul 10:41    INFO epoch 0 evaluating [time: 38.95s, valid_score: 0.054100]
21 Jul 10:41    INFO valid result:
recall@5: 0.0227    recall@10: 0.0354    recall@20: 0.0541    recall@50: 0.0909    ndcg@5: 0.0150    ndcg@10: 0.0192    ndcg@20: 0.0239    ndcg@50: 0.0313    precision@5: 0.0048    precision@10: 0.0038    precision@20: 0.0029    precision@50: 0.0020    map@5: 0.0123    map@10: 0.0139    map@20: 0.0152    map@50: 0.0164    mrr@5: 0.0130    mrr@10: 0.0148    mrr@20: 0.0162    mrr@50: 0.0174
21 Jul 10:41    INFO test result:
recall@5: 0.0224    recall@10: 0.0353    recall@20: 0.0538    recall@50: 0.0911    ndcg@5: 0.0150    ndcg@10: 0.0193    ndcg@20: 0.0240    ndcg@50: 0.0316    precision@5: 0.0050    precision@10: 0.0040    precision@20: 0.0030    precision@50: 0.0021    map@5: 0.0121    map@10: 0.0138    map@20: 0.0151    map@50: 0.0163    mrr@5: 0.0135    mrr@10: 0.0154    mrr@20: 0.0168    mrr@50: 0.0181
21 Jul 10:41    INFO ██ DRAGON--Best validation results updated!!!
```
We did not finish running but got one epoch. Maybe you can check your env setting and if correct use dataset, model, and config. 
I download ```DRAGON.py``` and ```dragon.yaml``` from this repo to run, so if you did not change these two files, the result should be what the above shows.

**Comment by zhongshsh** [comment]
Thanks a lot! I am currently unable to identify the cause of the error. My solution is to change https://github.com/enoche/MMRec/blob/master/src/models/dragon.py#L62 to :
```
mm_adj_file = os.path.join(dataset_path, 'mm_adj_{}.pt'.format(self.knn_k))
if config['dataset'] == 'elec':
    mm_adj_file = os.path.join(dataset_path, 'mm_adj_freedomdsp_10_1.pt')
```

**Comment by hongyurain** [comment]
The reason might be that the previous mm_adj file exists, so you are using an existing one for other models. 
You have already solved the problem, then I close this issue.
